File: src/store/status.ts
import { mkdir, rm } from "node:fs/promises"
import { dirname, join } from "node:path"
import { emitTicketEventFromEnv } from "../clawhip"
import { InvalidTransitionError, ReviewArtifactRequiredError } from "../errors"
import { type StatusInput, StatusInputSchema, type Ticket, type TicketStatus } from "../schema"
import { readActiveTicket, removeIndex, upsertIndex, writeJson, writeTicket } from "./io"
import { ensureStore, resolveStorePaths } from "./paths"
import { utcNow } from "./time"

export async function updateStatus(
  inputValue: StatusInput,
  paths = resolveStorePaths(),
): Promise<{
  readonly id: string
  readonly oldStatus: TicketStatus
  readonly newStatus: TicketStatus
}> {
  await ensureStore(paths)
  const input = StatusInputSchema.parse(inputValue)
  const ticket = await readActiveTicket(paths, input.id)
  if (!canTransition(ticket.status, input.status)) {
    throw new InvalidTransitionError(ticket.status, input.status)
  }
  if (input.status === "review" && input.artifact === undefined && ticket.artifacts.length === 0) {
    throw new ReviewArtifactRequiredError(ticket.id)
  }

  const now = utcNow()
  const oldStatus = ticket.status
  const nextTicket: Ticket = {
    ...ticket,
    status: input.status,
    updated: now,
    closed: input.status === "done" ? now : ticket.closed,
    artifacts: [...ticket.artifacts, ...artifactEntries(now, input.artifact, input.evidence)],
    log: [
      ...ticket.log,
      {
        ts: now,
        action: `${oldStatus} -> ${input.status}`,
        ...(input.note === undefined ? {} : { note: input.note }),
      },
    ],
  }

  if (input.status === "done") {
    const archiveFile = join(paths.archive, now.slice(0, 7), `${ticket.id}.json`)
    await mkdir(dirname(archiveFile), { recursive: true })
    await writeJson(archiveFile, nextTicket)
    await rm(join(paths.active, `${ticket.id}.json`))
    await removeIndex(paths, ticket.id)
  } else {
    await writeTicket(paths, nextTicket)
    await upsertIndex(paths, nextTicket)
  }

  await emitTicketEventFromEnv({
    kind: "ticket.status_changed",
    ticket: nextTicket,
    fromStatus: oldStatus,
    toStatus: input.status,
  })
  return { id: ticket.id, oldStatus, newStatus: input.status }
}

function artifactEntries(
  now: string,
  artifact?: string,
  evidence?: string,
): readonly Ticket["artifacts"][number][] {
  return [
    ...(artifact === undefined ? [] : [{ type: "artifact", value: artifact, ts: now }]),
    ...(evidence === undefined ? [] : [{ type: "evidence", value: evidence, ts: now }]),
  ]
}

function canTransition(from: TicketStatus, to: TicketStatus): boolean {
  switch (from) {
    case "open":
      return to === "doing" || to === "blocked"
    case "doing":
      return to === "review" || to === "blocked" || to === "open"
    case "review":
      return to === "done" || to === "open"
    case "blocked":
      return to === "doing" || to === "open"
    case "done":
      return false
  }
}

File: rust-ticket-flow/crates/ticket-flow-core/src/store/status.rs
use std::collections::BTreeMap;

use serde_json::json;

use super::TicketStore;
use crate::clawhip::{ClawhipEventKind, EnvClawhipEmit, emit_ticket_event_from_env};
use crate::error::{Result, TicketFlowError};
use crate::event::{EventEnvelope, TICKET_STATUS_CHANGED};
use crate::model::{Artifact, LogEntry, Ticket, TicketId, TicketStatus};
use crate::time::now_rfc3339;

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct StatusPatch {
    pub status: TicketStatus,
    pub artifact: Option<String>,
    pub evidence: Option<String>,
    pub note: Option<String>,
}

impl TicketStore {
    pub fn update_status(&self, id: &TicketId, patch: StatusPatch) -> Result<Ticket> {
        let mut ticket = self.load_active(id)?;
        let from = ticket.status;
        if !can_transition(from, patch.status) {
            return Err(TicketFlowError::InvalidTransition {
                from,
                to: patch.status,
            });
        }
        if patch.status == TicketStatus::Review
            && patch.artifact.is_none()
            && ticket.artifacts.is_empty()
        {
            return Err(TicketFlowError::ReviewArtifactRequired);
        }

        let now = now_rfc3339();
        ticket.status = patch.status;
        ticket.updated = now.clone();
        if patch.status == TicketStatus::Done {
            ticket.closed = Some(now.clone());
        }
        append_status_artifacts(&mut ticket, &patch, &now);
        ticket
            .log
            .push(status_log(from, patch.status, patch.note.clone(), now));

        let event = EventEnvelope::new(
            id.clone(),
            TICKET_STATUS_CHANGED,
            json!({ "from": from, "to": patch.status, "note": patch.note }),
        );
        if patch.status == TicketStatus::Done {
            self.archive_ticket(&ticket, event)?;
        } else {
            self.commit_active(&ticket, event)?;
        }
        emit_ticket_event_from_env(EnvClawhipEmit {
            kind: ClawhipEventKind::StatusChanged,
            ticket: &ticket,
            from_status: Some(from),
            to_status: Some(patch.status),
        });
        Ok(ticket)
    }
}

fn append_status_artifacts(ticket: &mut Ticket, patch: &StatusPatch, now: &str) {
    if let Some(value) = &patch.artifact {
        ticket
            .artifacts
            .push(Artifact::new("artifact", value.clone(), now.to_owned()));
    }
    if let Some(value) = &patch.evidence {
        ticket
            .artifacts
            .push(Artifact::new("evidence", value.clone(), now.to_owned()));
    }
}

fn status_log(from: TicketStatus, to: TicketStatus, note: Option<String>, now: String) -> LogEntry {
    LogEntry {
        ts: now,
        action: Some(format!("{from} -> {to}")),
        note,
        phase: None,
        decision: None,
        evidence: None,
        blocker: None,
        next: None,
        next_action: None,
        extra: BTreeMap::new(),
    }
}

const fn can_transition(from: TicketStatus, to: TicketStatus) -> bool {
    match from {
        TicketStatus::Open => matches!(to, TicketStatus::Doing | TicketStatus::Blocked),
        TicketStatus::Doing => matches!(
            to,
            TicketStatus::Review | TicketStatus::Blocked | TicketStatus::Open
        ),
        TicketStatus::Review => matches!(to, TicketStatus::Done | TicketStatus::Open),
        TicketStatus::Blocked => matches!(to, TicketStatus::Doing | TicketStatus::Open),
        TicketStatus::Done => false,
    }
}
